### PR TITLE
ppc pseudo

### PR DESCRIPTION
**Pull request by wargio**
support to standard PPC[32/64] pseudo + AltiVec (not all, but most yes).

emulation will come later, when i'll finish the basic support.

### CODE REVIEW

File: libr/parse/p/parse_ppc_pseudo.c [hunk 1]
```diff
@@ -10,231 +10,1460 @@
 #include <r_anal.h>
 #include <r_parse.h>
 
+ut64 mask64(ut64 mb, ut64 me) {
+    ut64 mask = 0;
```
Comment (radare) [inline libr/parse/p/parse_ppc_pseudo.c]: use tabs

File: libr/parse/p/parse_ppc_pseudo.c [hunk 2]
```diff
@@ -10,231 +10,1460 @@
...
+        mask = lo | hi;
+    }
+    return mask;
+}
+
+const char* cmask64(const char *mb_c, const char *me_c){
+    static char cmask[32];
+    ut64 mb = 0;
+    ut64 me = 0;
+    if(mb_c)
```
Comment (radare) [inline libr/parse/p/parse_ppc_pseudo.c]: always add spaces before ( (unless its a function definition)

File: libr/parse/p/parse_ppc_pseudo.c [hunk 3]
```diff
@@ -10,231 +10,1460 @@
...
+        mb = atol(mb_c);
+    if(me_c)
```
Comment (radare) [inline libr/parse/p/parse_ppc_pseudo.c]: add braces even for single-line blocks

File: libr/parse/p/parse_ppc_pseudo.c (hunk 1 above)
Comment (wargio) [inline libr/parse/p/parse_ppc_pseudo.c]: doh, i'm sorry. my editor changes them automatically..

File: libr/parse/p/parse_ppc_pseudo.c [hunk 4]
```diff
@@ -10,231 +10,1460 @@
...
+        me = atol(me_c);
+    snprintf(cmask, 32, "0x%"PFMT64x"", mask64(mb, me));
```
Comment (radare) [inline libr/parse/p/parse_ppc_pseudo.c]: use sizeof(cmask) instead of 32

File: libr/parse/p/parse_ppc_pseudo.c [hunk 5]
```diff
@@ -10,231 +10,1460 @@
...
+    return cmask;
+}
+
+ut32 mask32(ut32 mb, ut32 me) {
+    ut32 mask = 0;
+    if (mb < 0 || me < 0 || mb > 31 || me > 31) {
+        return mask;
+    } else if (mb < (me + 1)) {
+        for(int i = mb; i <= me ; i= i + 1) {
+            mask = mask | (ut32)(1LL<<(31-i));
```
Comment (radare) [inline libr/parse/p/parse_ppc_pseudo.c]: spaces between operators

File: libr/parse/p/parse_ppc_pseudo.c [hunk 6]
```diff
@@ -10,231 +10,1460 @@
...
+        me = atol(me_c);
+    snprintf(cmask, 32, "0x%"PFMT64x"", mask64(mb, me));
+    return cmask;
+}
+
+ut32 mask32(ut32 mb, ut32 me) {
+    ut32 mask = 0;
+    if (mb < 0 || me < 0 || mb > 31 || me > 31) {
+        return mask;
+    } else if (mb < (me + 1)) {
```
Comment (radare) [inline libr/parse/p/parse_ppc_pseudo.c]: dont  else after return

File: libr/parse/p/parse_ppc_pseudo.c [hunk 7]
```diff
@@ -10,231 +10,1460 @@
...
+//      { "xxsel", "A = B + C + D", 4},
+        { "xxsldwi", "A = B << C", 3},
+        { "xxspltd", "A = split(B)", 2},
+        { "xxspltw", "A = (word vector) split(B)", 2},
+        { "xxswapd", "swap(A,B)", 2},
+        { NULL }
+    };
+
+    for (i = 0; ops[i].op != NULL; i++) {
+        if (strcmp (ops[i].op, argv[0]) == 0) {
```
Comment (radare) [inline libr/parse/p/parse_ppc_pseudo.c]: use (!strcmp instead of == 0

File: libr/parse/p/parse_ppc_pseudo.c [hunk 8]
```diff
@@ -10,231 +10,1460 @@
...
+                            // If no parameters t[dw][i] 32, 0, 0 just TRAP
+                                w = "==";
+                            }
+                            w = cmask64("0", w);
+                        }
+                        if (w != NULL) {
+                            strcpy (newstr + k, w);
+                            k += strlen(w)-1;
+                        }
+                    } else newstr[k] = ops[i].str[j];
```
Comment (radare) [inline libr/parse/p/parse_ppc_pseudo.c]: add braces for this else block

File: libr/parse/p/parse_ppc_pseudo.c [hunk 9]
```diff
@@ -10,231 +10,1460 @@
...
+            return true;
+        }
+    }
+
+    /* TODO: this is slow */
+    if (newstr != NULL) {
+        newstr[0] = '\0';
+        for (i = 0; i < argc; i++) {
+            strcat (newstr, argv[i]);
+            strcat (newstr, (i == 0 || i == argc - 1) ? " ":", ");
```
Comment (radare) [inline libr/parse/p/parse_ppc_pseudo.c]: looks overflowable, we dont control the size of newstr here

File: libr/parse/p/parse_ppc_pseudo.c (hunk 8 above)
Comment (wargio) [inline libr/parse/p/parse_ppc_pseudo.c]: that wasn't touched, but ok.

File: libr/parse/p/parse_ppc_pseudo.c (hunk 9 above)
Comment (wargio) [inline libr/parse/p/parse_ppc_pseudo.c]: true, but i didn't change the code there, it was there before, but i'll add checks, for sure.

Comment (wargio) [inline libr/parse/p/parse_ppc_pseudo.c]: how do i check the length of that? the string pointer comes from the `parse(RParse *p, const char *data, char *str)` caller.

**Comment by radare** [comment]
we cant, thats probably a design issue because i assume the resultng string will be smaller than the input one. but i guess passing another argument to specify the length will be good here

**Comment by wargio** [comment]
totally, from my point of view. there are some instructions that are quite complex on ppc arch

**Comment by radare** [comment]
Do you wanna do this change? i know that's out of the scope of this PR, but if you are on the mood. it will be something good to have. if not just fill an issue to track this out

**Comment by radare** [comment]
Not completelly following the coding style, but enough to merge. Travis is green again :)

**Comment by wargio** [comment]
what you mean? anyway, i'm going to push 2 more things, because i found 2 errors.
My coding style was a mixup of javascript "pre-testing" imported to C later.
I was doing that and my editor keeps reformatting things, that's why.

**Comment by radare** [comment]
Ok i can merge those fixes later. And ill fix the indentation later. Dont worry.

Im talking about doing the change that adds the length parameter to this function delegate

**Comment by wargio** [comment]
Perfect! thanks! :)
i might work on that when i'll finish with the ppc emu

**Pull request merged**